bundles/org.openhab.ui.homebuilder/web/src/formSchema.js

```diff
@@ -0,0 +1,275 @@
+            .trim()
+            .toLowerCase()
+            .cleanDiacritics()
+            .classify()
+            .value()
+    }
+    rooms.push(tag);
+    value.push(tag);
+};
+
+/**
+ * Is being executed when 
+ * collection of rooms in floor multiselect field
+ * has changed.
+ * 
+ * If there's a new `room` in collection,
+ * a new dynamic field is added to the floor object, e.g.
+ * `"GroundFloor": [ { name: 'Bedroom', value: 'Bedroom', icon: 'bed' }]`
+ * 
+ * If an entry was removed from the collection,
+ * a dynamic field is removed accordingly.
+ * 
+ * @param {Object} model 
+ * @param {Array} newVal
+ * @param {Array} oldVal 
+ * @param {Object} field 
+ */
+function roomsChanged(model, newVal, oldVal, field) {
+    let devicesFields = this.$options.parent.groups[3].fields;
+    let floor = field.model;
+    let oldList = oldVal ? oldVal.map((room) => room.value) : [];
+    let newList = newVal.map((room) => room.value);
+    let lastRemoved = _.first(_.difference(oldList, newList));
+    let lastItem = _.last(newList);
+    let roomName = '';
+
+    if (lastItem && !lastRemoved) {
+        let room = _.find(rooms, { value: lastItem });
+        roomName = floor + '_' + room.value + DEVICES_SUFFIX;
+        devicesFields.push({
+            type: 'multiselect',
+            label: (room.name || room.value) + ' (' + _.find(floors, { value: floor }).name + ')',
+            styleClasses: 'rooms-list',
+            model: roomName,
+            placeholder: 'Type to search device',
+            selectOptions: {
+                multiple: true,
+                hideSelected: true,
+                closeOnSelect: false,
+                selectLabel: '',
+                trackBy: 'value',
+                label: 'name',
+                searchable: true,
+                taggable: false
+            },
+            values: devices
+        });
+    }
+
+    if (lastRemoved) {
+        roomName = floor + '_' + lastRemoved + DEVICES_SUFFIX;
+        _.remove(devicesFields, { model: roomName });
+        delete model[roomName];
+    }
+}
+
+/**
+ * Schema describing the basic form
+ * generated by vue-form-generator
+ */
+export var basicFields = [{
+        type: 'select',
+        model: 'language',
+        label: 'Please select your language',
+        values: function() {
+            return languages;
+        },
+        selectOptions: {
+            hideNoneSelectedText: true
+        },
+        onChanged: languageChanged
+    },
+
+    {
+        type: 'input',
+        inputType: 'text',
+        label: 'Home Setup Name',
+        model: 'homeName',
+        readonly: false,
+        disabled: false,
+        placeholder: 'Home name',
+        validator: validators.string
+    }
+];
+
+export var floorsFields = [{
+    type: 'radios',
+    model: 'floorsCount',
+    label: 'Number of floors',
+    styleClasses: 'floors-number',
+    values: [1, 2, 3, 4, 5],
+    onChanged: function(model, newVal, oldVal, field) {
```

Comment:
pass by reference, refactor complexity

Reply:
[it's a part of API](https://icebob.gitbooks.io/vueformgenerator/content/fields/field_properties.html#common-events-of-field), so not much I can do here.